core/trino-main/src/main/java/io/trino/server/protocol/Query.java

```diff
@@ -459,18 +459,24 @@ private synchronized QueryResults getNextResult(long token, UriInfo uriInfo, Dat
         QueryInfo queryInfo = queryManager.getFullQueryInfo(queryId);
         queryManager.recordHeartbeat(queryId);
 
-        closeExchangeIfNecessary(queryInfo);
```

Comment:
I think `closeExchangeIfNecessary(QueryInfo)` needs to check `queryInfo.getState().ordinal() > STARTING` logic too, otherwise we could close the exchange before planning completes if that were to be called from another code path.

After that change, we should leave the `closeExchangeIfNecessary` and `removePagesFromExchange` as-is without needing to check `isStarted` at this point explicitly.